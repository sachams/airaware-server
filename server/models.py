from sqlalchemy import Boolean, Column, ForeignKey, Integer, String, Float, DateTime
from sqlalchemy.orm import relationship

from .database import Base


class SensorData(Base):
    __tablename__ = "sensor_data"

    id = Column(Integer, primary_key=True, index=True)
    site_code = Column(String, index=True)
    series = Column(String, index=True)
    time = Column(DateTime(timezone=True))
    value = Column(Float)
