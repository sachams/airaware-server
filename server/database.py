from app_config import database_host, database_name, database_password, database_username
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

SQLALCHEMY_DATABASE_URL = (
    f"postgresql://{database_username}:{database_password}@{database_host}:"
    f"5432/{database_name}?sslmode=disable"
)

engine = create_engine(SQLALCHEMY_DATABASE_URL)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
