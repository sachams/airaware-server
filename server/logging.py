import logging


def configure_logging():
    logging.basicConfig(format="%(asctime)s %(levelname)s %(message)s", level=logging.INFO)
    logger = logging.getLogger("httpx")
    logger.setLevel(logging.WARNING)
