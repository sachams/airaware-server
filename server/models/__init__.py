from server.models.broken_site_model import BrokenSiteModel
from server.models.request_log_model import RequestLogModel
from server.models.site_model import SiteModel
from server.models.sensor_data_model import SensorDataModel


__all__ = [
    "BrokenSiteModel",
    "RequestLogModel",
    "SiteModel",
    "SensorDataModel",
]
