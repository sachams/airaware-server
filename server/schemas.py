import datetime
from pydantic import BaseModel


class SensorDataBase(BaseModel):
    time: datetime.datetime
    value: float


class SensorDataCreate(SensorDataBase):
    pass


class SensorData(SensorDataBase):
    class Config:
        orm_mode = True


class SiteAverage(BaseModel):
    site_code: str
    value: float


class RequestLogBase(BaseModel):
    time: datetime.datetime
    ip_address: str
    path: str


class RequestLogCreate(RequestLogBase):
    pass


class RequestLog(RequestLogBase):
    id: int

    class Config:
        orm_mode = True
