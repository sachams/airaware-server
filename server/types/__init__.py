from server.types.classification import Classification
from server.types.data_status import DataStatus
from server.types.frequency import Frequency
from server.types.series import Series
from server.types.site_status import SiteStatus
from server.types.source import Source

__all__ = [
    "Classification",
    "DataStatus",
    "Frequency",
    "Series",
    "SiteStatus",
    "Source",
]
