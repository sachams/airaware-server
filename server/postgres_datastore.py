import logging
import pyarrow as pa
import datetime
from sqlalchemy import func
from sqlalchemy.orm import Session, aliased
from sqlalchemy import desc, select

from . import models, schemas


class PostgresDatastore:
    def __init__(self, db: Session):
        self.db = db

    def write_request_log(self, url: str, ip_address: str):
        """Records requests"""
        db_item = models.RequestLog(
            path=url, ip_address=ip_address, time=datetime.datetime.utcnow()
        )
        self.db.add(db_item)
        self.db.commit()

    def write_data(self, site_code, series, data: list[schemas.SensorDataCreate]):
        """Writes Breathe London series data to the database"""
        objects = []
        for item in data:
            db_item = models.SensorData(
                site_code=site_code, series=series, time=item.time, value=item.value
            )
            objects.append(db_item)
        self.db.bulk_save_objects(objects)
        self.db.commit()

    def read_data(self, series, start, end, site_codes, frequency):
        """Reads data from the datastore, averaging across the specified sites. If no
        sites are specified, it averages across all sites.

        site_codes: a list of sites to average over. If empty, it averages over all
                    sites.
        frequency:  either `hourly` or `daily`
        """
        frequency_mapper = {"hourly": "hour", "daily": "day"}
        db_frequency = frequency_mapper.get(frequency)
        if db_frequency is None:
            raise Exception(f"Unknown frequency {frequency}")

        query = (
            select(
                func.date_trunc(db_frequency, models.SensorData.time).label("time"),
                func.avg(models.SensorData.value).label("value"),
            )
            .filter(models.SensorData.series == series)
            .filter(models.SensorData.time >= start)
            .filter(models.SensorData.time < end)
        )

        if site_codes:
            query = query.filter(models.SensorData.site_code.in_(site_codes))

        query = query.group_by(
            func.date_trunc(db_frequency, models.SensorData.time).label("time")
        ).order_by(func.date_trunc(db_frequency, models.SensorData.time))

        return self.db.execute(query)

    def read_site_average(self, series, start, end) -> list[tuple[str, float]]:
        """Reads data from the datastore, returning the average of all sites
        across the specified time period. Data is returned as list of site_code
        and average value.
        """
        return self.db.execute(
            select(
                models.SensorData.site_code,
                func.avg(models.SensorData.value).label("value"),
            )
            .filter(models.SensorData.series == series)
            .filter(models.SensorData.time >= start)
            .filter(models.SensorData.time < end)
            .group_by(models.SensorData.site_code)
            .order_by(models.SensorData.site_code)
        )

    def get_latest_date(self, site_code, series) -> datetime.datetime:
        reading = (
            self.db.execute(
                select(models.SensorData)
                .filter(models.SensorData.site_code == site_code)
                .filter(models.SensorData.series == series)
                .order_by(desc(models.SensorData.time))
                .limit(1)
            )
            .scalars()
            .first()
        )

        return None if reading is None else reading.time
