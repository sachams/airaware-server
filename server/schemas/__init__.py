from server.schemas.request_log_schema import RequestLogSchema
from server.schemas.sensor_data_schema import (
    SensorDataCreateSchema,
    SensorDataRemoteSchema,
    SensorDataSchema,
)
from server.schemas.site_average_schema import SiteAverageSchema
from server.schemas.site_schema import SiteCreateSchema, SiteSchema

__all__ = [
    "RequestLogSchema",
    "SensorDataCreateSchema",
    "SensorDataRemoteSchema",
    "SensorDataSchema",
    "SiteAverageSchema",
    "SiteCreateSchema",
    "SiteSchema",
]
