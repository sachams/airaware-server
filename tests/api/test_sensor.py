from http import HTTPStatus

import pytest
from sqlalchemy.exc import DatabaseError

from server.types import Classification


@pytest.mark.usefixtures("use_fake_uow")
def test_get_data(client, snapshot):
    response = client.get("/sensor/pm25/2022-01-01T10:00:00/2022-02-01T10:00:00/hour")
    assert response.status_code == HTTPStatus.OK
    assert response.json() == snapshot


@pytest.mark.usefixtures("use_fake_uow")
def test_get_data_site_code(client, snapshot):
    response = client.get(
        "/sensor/pm25/2022-01-01T10:00:00/2022-02-01T10:00:00/hour?codes=CLDP0002"
    )
    assert response.status_code == HTTPStatus.OK
    assert response.json() == snapshot


@pytest.mark.usefixtures("use_fake_uow")
def test_get_data_type(client, snapshot):
    response = client.get(
        f"/sensor/pm25/2022-01-01T10:00:00/2022-02-01T10:00:00/hour?types={Classification.roadside}"
    )
    assert response.status_code == HTTPStatus.OK
    assert response.json() == snapshot


@pytest.mark.usefixtures("use_fake_uow")
def test_get_site_average(client, snapshot):
    response = client.get("/site_average/pm25/2022-01-01T10:00:00/2022-02-01T10:00:00")
    assert response.status_code == HTTPStatus.OK
    assert response.json() == snapshot


@pytest.mark.usefixtures("use_fake_uow")
def test_get_sites(client, snapshot):
    response = client.get("/sites")
    assert response.status_code == HTTPStatus.OK
    assert response.json() == snapshot


@pytest.mark.usefixtures("use_fake_uow")
def test_get_site_average_db_exception(client, mocker, sensor_repository):
    sensor_repository.get_site_average = mocker.MagicMock(
        side_effect=DatabaseError(params=None, orig=None, statement="Test exception")
    )

    response = client.get("/site_average/pm25/2022-01-01T10:00:00/2022-02-01T10:00:00")
    assert response.status_code == HTTPStatus.INTERNAL_SERVER_ERROR
    assert response.status_code == HTTPStatus.INTERNAL_SERVER_ERROR
