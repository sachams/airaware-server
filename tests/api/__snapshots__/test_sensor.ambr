# serializer version: 1
# name: test_get_data
  list([
    dict({
      'time': '2020-06-05T03:02:01',
      'value': 1.23,
    }),
    dict({
      'time': '2020-06-05T03:02:01',
      'value': 1.23,
    }),
  ])
# ---
# name: test_get_site_average
  list([
    dict({
      'site_code': 'CLDP0001',
      'value': 1.23,
    }),
    dict({
      'site_code': 'CLDP0002',
      'value': 4.56,
    }),
  ])
# ---
# name: test_get_sites
  list([
    dict({
      'borough': 'Lambeth',
      'description': 'The Royal London Hospital is home to one of the largest...',
      'end_date': '2022-06-05T03:02:01',
      'latitude': 51.518775939941406,
      'longitude': -0.059463899582624435,
      'name': 'Royal London University Hospital',
      'photo_url': 'https://api.breathelondon.org/assets/images/CLDP0001.jpg',
      'site_code': 'CLDP0001',
      'site_id': 1,
      'site_type': 'urban_background',
      'source': 'breathe_london',
      'start_date': '2020-06-05T03:02:01',
      'status': 'healthy',
    }),
    dict({
      'borough': 'Lambeth',
      'description': 'Brixton Central',
      'end_date': None,
      'latitude': 50.0,
      'longitude': -0.01,
      'name': 'Brixton',
      'photo_url': 'https://api.breathelondon.org/assets/images/CLDP0002.jpg',
      'site_code': 'CLDP0002',
      'site_id': 2,
      'site_type': 'roadside',
      'source': 'breathe_london',
      'start_date': '2022-06-05T03:02:01',
      'status': 'coming_online',
    }),
  ])
# ---
